chrome/content/zotero/xpcom/connector/server_connector.js

```diff
@@ -1947,3 +1971,50 @@ Zotero.Server.Connector.Request.prototype = {
 		})];
 	}
 };
+
+Zotero.Server.Connector.Cancel = function() {};
+Zotero.Server.Endpoints["/connector/cancel"] = Zotero.Server.Connector.Cancel;
+Zotero.Server.Connector.Cancel.prototype = {
+	supportedMethods: ["POST"],
+	supportedDataTypes: ["application/json"],
+	
+	init: async function (request) {
+		let { sessionID } = request.data;
+		var session = Zotero.Server.Connector.SessionManager.get(sessionID);
+
+		// Session might not exist if the cancel button is clicked before any of /connector/save*
+		// endpoints are invoked, in which case the session is never created, so there's nothing to cancel
+		if (!session) {
+			return [200, "application/json", JSON.stringify({ status: "NO_SESSION" })];
+		}
+
+		session.cancelled = true;
+		// If all items have been saved, delete them now
+		if (session.isSavingDone()) {
+			await session.eraseItems();
+			return [200, "application/json", JSON.stringify({ status: "CANCELLED" })];
+		}
+		
+		// Otherwise, trash them so they go immediately away and begin checking if the saving is finished.
+		// As soon as it is, erase all items from this session.
+		await session.trashItems();
+		let checkCount = 0;
+		const intervalDelay = 100;
+		const maxMinutes = 3;
+		let intervalID = setInterval(async () => {
+			// In case any additional top-level items are added, make sure they are trashed too
+			await session.trashItems();
+			if (session.isSavingDone()) {
+				session.eraseItems();
+				clearInterval(intervalID);
+			}
+			// Ensure that even if something goes wrong, the loop breaks after 3 minutes
+			checkCount += 1;
+			if (checkCount > maxMinutes * 60 * 1000 / intervalDelay) {
+				clearInterval(intervalID);
+			}
+		}, intervalDelay);
```

Comment:
Move this into the session and make it return a promise, then in the test spy the function and await the returned promise so that we don't have to add more arbitrary delays than necessary.

Reply:
I moved it all into `session.eraseItemsWhenDone`, which the second test can `await` for. It's definitely cleaner that way!

I'm not sure if we actually need two different tests to check for item deletion now. They are handled differently but ultimately, the call to `/connector/cancel` when the save process is finished is like a subset of what is being done if `/connector/cancel` is called when the save process is ongoing. So maybe I should drop the `should delete saved items if called after everything is saved` test?

Reply:
> So maybe I should drop the `should delete saved items if called after everything is saved` test?

No harm in having it. Let's keep it in case something changes in the future.